File: app/src/main/java/com/app/memoeslink/adivinador/activity/SettingsActivity.java
package com.app.memoeslink.adivinador.activity;

import android.content.SharedPreferences;
import android.media.AudioManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.speech.tts.TextToSpeech;
import android.view.KeyEvent;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;

import com.app.memoeslink.adivinador.R;
import com.app.memoeslink.adivinador.Screen;
import com.app.memoeslink.adivinador.preference.Preference;
import com.app.memoeslink.adivinador.preference.PreferenceHandler;
import com.takisoft.preferencex.PreferenceFragmentCompat;

public class SettingsActivity extends CommonActivity implements TextToSpeech.OnInitListener {
    private AlertDialog dialog;
    private AudioManager audioManager;
    private TextToSpeech tts;
    private SharedPreferences.OnSharedPreferenceChangeListener listener; //Declared as global to avoid destruction by JVM Garbage Collector

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportFragmentManager().beginTransaction().replace(android.R.id.content, new CustomPreferenceFragment()).commit();
        audioManager = (AudioManager) getSystemService(AUDIO_SERVICE);
        tts = new TextToSpeech(this, this);

        AlertDialog.Builder builder = new AlertDialog.Builder(SettingsActivity.this);
        builder.setTitle(getString(R.string.alert_app_restart_title));
        builder.setMessage(getString(R.string.alert_app_restart_message));
        builder.setIcon(R.drawable.exit);
        builder.setNeutralButton(getString(R.string.ok), (dialog, which) -> dialog.dismiss());
        builder.setOnDismissListener(arg0 -> restartApplication());
        dialog = builder.create();

        //Set listeners
        listener = (prefs, key) -> {
            if (key.equals(Preference.SETTING_ACTIVE_SCREEN.getTag()))
                Screen.setContinuance(SettingsActivity.this, PreferenceHandler.getBoolean(Preference.SETTING_ACTIVE_SCREEN));

            if (key.equals(Preference.SETTING_ADS_ENABLED.getTag()) && PreferenceHandler.getBoolean(Preference.SETTING_ADS_ENABLED))
                PreferenceHandler.put(Preference.TEMP_RESTART_ACTIVITY, true);

            if (key.equals(Preference.SETTING_FORTUNE_TELLER_ASPECT.getTag()))
                PreferenceHandler.put(Preference.TEMP_CHANGE_FORTUNE_TELLER, true);

            if (key.equals(Preference.SETTING_LANGUAGE.getTag()) || key.equals(Preference.SETTING_THEME.getTag()) || key.equals(Preference.SETTING_SEED.getTag())) {
                try {
                    SettingsActivity.this.runOnUiThread(() -> dialog.show());
                } catch (Exception e) {
                    new Handler(Looper.getMainLooper()).postDelayed(this::restartApplication, 500);
                }
            }

            if (key.equals(Preference.SETTING_SAVE_NAMES.getTag()) && !PreferenceHandler.getBoolean(Preference.SETTING_SAVE_NAMES, true))
                PreferenceHandler.remove(Preference.DATA_STORED_NAMES);

            if (key.equals(Preference.SETTING_SAVE_ENQUIRIES.getTag()) && !PreferenceHandler.getBoolean(Preference.SETTING_SAVE_ENQUIRIES, true))
                PreferenceHandler.remove(Preference.DATA_STORED_PEOPLE);
        };
        PreferenceHandler.changeDefaultPreferencesListener(listener);
    }

    @Override
    public void onResume() {
        super.onResume();
        Screen.setContinuance(SettingsActivity.this, PreferenceHandler.getBoolean(Preference.SETTING_ACTIVE_SCREEN));
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return true;
    }

    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        int action = event.getAction();
        int keyCode = event.getKeyCode();

        switch (keyCode) {
            case KeyEvent.KEYCODE_VOLUME_UP:
                if (action == KeyEvent.ACTION_DOWN) {
                    if (PreferenceHandler.getBoolean(Preference.SETTING_AUDIO_ENABLED))
                        audioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_RAISE, AudioManager.FLAG_SHOW_UI);
                    else
                        audioManager.adjustStreamVolume(AudioManager.STREAM_RING, AudioManager.ADJUST_RAISE, AudioManager.FLAG_SHOW_UI);
                }
                return true;
            case KeyEvent.KEYCODE_VOLUME_DOWN:
                if (action == KeyEvent.ACTION_DOWN) {
                    if (PreferenceHandler.getBoolean(Preference.SETTING_AUDIO_ENABLED))
                        audioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_LOWER, AudioManager.FLAG_SHOW_UI);
                    else
                        audioManager.adjustStreamVolume(AudioManager.STREAM_RING, AudioManager.ADJUST_LOWER, AudioManager.FLAG_SHOW_UI);
                }
                return true;
            default:
                return super.dispatchKeyEvent(event);
        }
    }

    @Override
    public void onInit(int i) {
        if (i == TextToSpeech.SUCCESS)
            speechAvailable = isTTSAvailable(tts, PreferenceHandler.getString(Preference.SETTING_LANGUAGE, "es"));
        else {
            speechAvailable = false;
            showToast(getString(R.string.toast_voice_unavailability));
        }
    }

    public static class CustomPreferenceFragment extends PreferenceFragmentCompat {
        @Override
        public void onCreatePreferencesFix(@Nullable Bundle savedInstanceState, String rootKey) {
            setPreferencesFromResource(R.xml.default_preferences, rootKey);
        }
    }
}

File: app/src/main/java/com/app/memoeslink/adivinador/activity/SettingsActivity.kt
package com.app.memoeslink.adivinador.activity

import android.content.DialogInterface
import android.content.SharedPreferences
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.preference.PreferenceFragmentCompat
import com.app.memoeslink.adivinador.R
import com.app.memoeslink.adivinador.Speech
import com.app.memoeslink.adivinador.extensions.setContinuance
import com.app.memoeslink.adivinador.preference.Preference
import com.app.memoeslink.adivinador.preference.PreferenceHandler

class SettingsActivity : CommonActivity() {
    private lateinit var dialog: AlertDialog
    private lateinit var listener: SharedPreferences.OnSharedPreferenceChangeListener // Declared as global to avoid destruction by JVM Garbage Collector

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportFragmentManager
            .beginTransaction()
            .replace(androidx.appcompat.R.id.content, CustomPreferenceFragment())
            .commit()

        dialog = AlertDialog.Builder(this@SettingsActivity)
            .setTitle(getString(R.string.alert_app_restart_title))
            .setMessage(getString(R.string.alert_app_restart_message)).setIcon(R.drawable.exit)
            .setNeutralButton(getString(R.string.ok)) { dialog: DialogInterface, _: Int -> dialog.dismiss() }
            .setOnDismissListener { restartApplication() }.create()

        listener = SharedPreferences.OnSharedPreferenceChangeListener { _, key ->
            when (key) {
                Preference.SETTING_AUDIO_ENABLED.tag, Preference.SETTING_VOICE_ENABLED.tag -> {
                    if (!PreferenceHandler.getBoolean(Preference.SETTING_AUDIO_ENABLED) || !PreferenceHandler.getBoolean(
                            Preference.SETTING_VOICE_ENABLED
                        )
                    ) Speech.getInstance(this@SettingsActivity).suppress()
                }

                Preference.SETTING_ACTIVE_SCREEN.tag -> this@SettingsActivity.setContinuance(
                    PreferenceHandler.getBoolean(Preference.SETTING_ACTIVE_SCREEN)
                )

                Preference.SETTING_ADS_ENABLED.tag -> PreferenceHandler.put(
                    Preference.TEMP_RESTART_ACTIVITY, true
                )

                Preference.SETTING_FORTUNE_TELLER_ASPECT.tag -> PreferenceHandler.put(
                    Preference.TEMP_CHANGE_FORTUNE_TELLER, true
                )

                Preference.SETTING_LANGUAGE.tag, Preference.SETTING_THEME.tag, Preference.SETTING_SEED.tag -> {
                    runOnUiThread {
                        dialog.show()
                    }
                }

                Preference.SETTING_SAVE_NAMES.tag -> {
                    if (!PreferenceHandler.getBoolean(Preference.SETTING_SAVE_NAMES)) PreferenceHandler.remove(
                        Preference.DATA_STORED_NAMES
                    )
                }

                Preference.SETTING_SAVE_ENQUIRIES.tag -> {
                    if (!PreferenceHandler.getBoolean(Preference.SETTING_SAVE_ENQUIRIES)) PreferenceHandler.remove(
                        Preference.DATA_STORED_PEOPLE
                    )
                }
            }
        }
        PreferenceHandler.changeDefaultPreferencesListener(listener)
    }

    override fun onResume() {
        super.onResume()
        this@SettingsActivity.setContinuance(PreferenceHandler.getBoolean(Preference.SETTING_ACTIVE_SCREEN))
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }

    class CustomPreferenceFragment : PreferenceFragmentCompat() {
        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            setPreferencesFromResource(R.xml.default_preferences, rootKey)
        }
    }
}
